### PR TITLE
Add kiali URL in traffic connector side panel

### PR DESCRIPTION
**Pull request by divyanshiGupta**
Jira Stories: https://issues.redhat.com/browse/ODC-2603, https://issues.redhat.com/browse/ODC-2569

kiali

Dependent on -

- https://github.com/openshift/console/pull/4022
- https://github.com/openshift/console/pull/4003

To review go through - https://github.com/openshift/console/commit/e3b5c5dc1e0f11c6e71f4e154f1ad53df5c655d5

### CODE REVIEW

**Comment by christianvogt** [comment]
@divyanshiGupta have you seen the ConsoleLink updates coming in from kiali for creating a namespaced link from the dashboard?
https://github.com/kiali/kiali/pull/2078/files
Constructing the URL as you do isn't supported as API by kiali, while the ConsoleLink will be a valid URL. How different are the landing pages between the two urls? If it doesn't take us where we need to go, then constructing for now is ok as we work through the integration issues with using their internal apis.

**Comment by divyanshiGupta** [comment]
> @divyanshiGupta have you seen the ConsoleLink updates coming in from kiali for creating a namespaced link from the dashboard?
> https://github.com/kiali/kiali/pull/2078/files

I havent seen the updates since I wanted to get everything in place first and then according to ConsoleLink updates make changes to the PR. 

> Constructing the URL as you do isn't supported as API by kiali, while the ConsoleLink will be a valid URL. How different are the landing pages between the two urls? If it doesn't take us where we need to go, then constructing for now is ok as we work through the integration issues with using their internal apis.

I will confirm

**Comment by divyanshiGupta** [comment]
@christianvogt both URLs will take us to the same page. So I will directly use only the `consoleLink`.

**Comment by christianvogt** [comment]
Thanks @divyanshiGupta.
My concern is that their URLs aren't officially considered api for external manipulation. That being said it's a URL and bookmarks would break if they did change. So really all URLs should be considered API. But i can see their argument against.
So using the console link urls is an api they buy into.

Seems you're doing a lot of work to weave the console links data through topology. Why don't you simply connect to redux in the `TopologyEdgePanel` and create the url on the spot using the console links and namespace from redux?

**Comment by divyanshiGupta** [comment]
> Seems you're doing a lot of work to weave the console links data through topology. Why don't you simply connect to redux in the `TopologyEdgePanel` and create the url on the spot using the console links and namespace from redux?

Because TopologyDataController is already a connected component and also I thought of passing the console-links from here since most of the data we need anywhere in topology is passed/controlled by this component. Just didn't wanted to break the flow. Definitely making `TopologyEdgePanel` a connected component and getting the console-links would have been straightforward. What do you suggest? Also we don't need namespace now since looking at the PR you shared they are going to provide namespaced URL so we can directly use consoleLink URL.

File: frontend/packages/dev-console/src/components/topology/topology-utils.ts [hunk 1]
```diff
@@ -348,6 +351,46 @@ export const getTopologyGroupItems = (dc: K8sResourceKind, groups: Group[]): Gro
...
+    );
+    const sourceResourceNode = resources.find((res) => {
+      return res.metadata.name === sourceTrafficNode[sourceTrafficNode.nodeType];
+    });
+    const targetResourceNode = resources.find(
+      (res) => res.metadata.name === targetTrafficNode[targetTrafficNode.nodeType],
+    );
+    const kialiLink = consoleLinks?.find(
+      (cl) => cl.metadata.name === `kiali-namespace-${sourceResourceNode?.metadata?.namespace}`,
+    )?.spec?.href;
```
Comment (karthikjeeyar) [inline frontend/packages/dev-console/src/components/topology/topology-utils.ts]: There will be one kiali link per namespace, we can avoid finding it multiple times.

File: frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import {
+  TYPE_CONNECTS_TO,
+  TYPE_EVENT_SOURCE_LINK,
+  TYPE_REVISION_TRAFFIC,
+  TYPE_SERVICE_BINDING,
+  TYPE_TRAFFIC_CONNECTOR,
+} from './const';
+import { edgeActions } from './actions/edgeActions';
+import * as classNames from 'classnames';
+import * as _ from 'lodash';
```
[the remark this answers is not in the document]
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx]: These are not my changes. I have added the commit in my PR description that has my code https://github.com/openshift/console/commit/a89ae368f3abb1c3e8d1a0ef2f5e5fd2b3a65808

File: frontend/packages/dev-console/src/components/topology/topology-utils.ts [hunk 3]
```diff
@@ -348,6 +351,46 @@ export const getTopologyGroupItems = (dc: K8sResourceKind, groups: Group[]): Gro
   return groups;
 };
 
+export const getFilteredTrafficWorkload = (nodes: KialiNode[]): KialiNode[] =>
+  nodes.filter(({ data }) => data.nodeType === 'workload');
```
[the remark this answers is not in the document]
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/topology-utils.ts]: These are not my changes. I have added the commit in my PR description that has my code a89ae36

**Comment by karthikjeeyar** [comment]
verified the changes in https://github.com/openshift/console/commit/e3b5c5dc1e0f11c6e71f4e154f1ad53df5c655d5
/lgtm

File: frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx [hunk 4]
```diff
@@ -1,26 +1,35 @@
...
+interface StateProps {
+  consoleLinks?: K8sResourceKind[];
+}
 
-export type TopologyEdgePanelProps = {
+export interface TopologyEdgePanelProps extends StateProps {
   edge: BaseEdge;
   data: TopologyDataModel;
-};
+}
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx]: You can extend `type`s just like `interface`s... the syntax just changes:

```
type StateProps = {};
type TopologyEdgePanelProps = {} & StateProps;
```

File: frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx [hunk 5]
```diff
@@ -32,17 +41,25 @@ const connectorTypeToTitle = (type: string): string => {
...
       return 'Event source connector';
+    case TYPE_TRAFFIC_CONNECTOR:
+      return 'Traffic connector';
     default:
       return '';
   }
 };
 
-const TopologyEdgePanel: React.FC<TopologyEdgePanelProps> = ({ edge, data }) => {
+export const TopologyEdgePanel: React.FC<TopologyEdgePanelProps> = ({
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx]: Is there any reason we want to export this on its own? I believe no one uses it but the one spot you are using the export default.

File: frontend/packages/dev-console/src/components/topology/topology-types.ts [hunk 6]
```diff
@@ -48,7 +48,7 @@ export interface Edge {
   type?: string;
   source: string;
   target: string;
-  data?: { sbr?: K8sResourceKind };
+  data?: {};
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/topology/topology-types.ts]: Why was this change needed? Am I missing something about your change that makes `sbr` (which I assume is service binding request) not be needed?

Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/topology-types.ts]: Its not needed here because this data part is variable as data can be different for all edge types.

File: frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx (hunk 5 above)
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx]: Yeah will remove it but later for testing it might be used elsewhere

File: frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx (hunk 4 above)
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx]: Yeah but in most of the places we use interface so for consistency across files I used interface will change it to type

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx]: I think interfaces were a poor knowledge decision originally. TypeScript has gone through a lot of changes... back in the day `type` wasn't available. Most of our stuff was classes before I believe, which implies `interfaces`. I don't think it really matters one way or another... but I favour `types` for all things type related and `interfaces` for class-based implementation... a la Java Interfaces.

Personally, I wouldn't change the type when modifying code, but either way, it's a minor nit.

File: frontend/packages/dev-console/src/components/topology/topology-types.ts (hunk 6 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/topology/topology-types.ts]: 🤔 I suppose that's fair. Feels like we are losing something here and not replacing it with a better solution.

File: frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx (hunk 4 above)
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/topology/TopologyEdgePanel.tsx]: Got it

**Comment by andrewballantyne** [comment]
/hold cancel

I'm going to throw all the chips on the table... If #4022 (the 1st commit of this PR & a dependency for this work) merges first, this PR will get conflicts and cannot merge... however if this merges before #4022 then 4022 won't be needed. We are having e2e failures that are blocking things and Sahil (author of 4022) will be on PTO during the next couple days preventing any response to failures / rebases / blockers that arise on #4022.

So, let the race begin 😄
